### PR TITLE
Separate Node-Based Graph Creation from Edge-Based Graph Factory

### PR DESCRIPTION
**Pull request by MoKob**
# Issue

Passing data from OSM through OSRM to the API requires the adjustment of a large list of files. This PR simplifies this process by using the NodeDataContainer in both pre and post processing.
In addition, it separates the creation of the node-based graph logically from the edge-based graph, letting the edge-based graph factory take care of it's mappings, not polluting the node-based graph edges.

Resolves https://github.com/Project-OSRM/osrm-backend/issues/2052

## Tasklist
 - [x] add changelog
 - [x] ticket further memory optimisations
 - [x] make performance measurements to analyse annotation speed
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

**Comment by MoKob** [comment]
extract

At the moment, there is a regression in memory-requirements for the resident-set size. So I will have to go through the code again and check where this (unexpected) increase comes from.

**Comment by MoKob** [comment]
The memory consumption issues have been resolved (held a copy of the node-based graph).

extractor

There are a few items that allow further optimisation. Right now the node-based edges hold their own geometry ID. These could be stored directly into the node-based edge container, instead of holding a copy for each of them.
Right now, the memory consumption is fine (no changes, even though we duplicate data), so I would argue to go ahead as is and ticket further optimisations for later.

File: include/extractor/node_based_edge.hpp [hunk 1]
```diff
@@ -12,105 +15,124 @@ namespace osrm
...
+{
+    NameID name_id;                        // 32 4
+    TravelMode travel_mode : 4;            // 4
+    ClassData classes;                     // 8  1
+    LaneDescriptionID lane_description_id; // 16 2
+    bool is_left_hand_driving;             // 1
+
+    bool CanCombineWith(const NodeBasedEdgeAnnotation &other) const
+    {
+        return (std::tie(name_id, classes) == std::tie(other.name_id, other.classes)) &&
```
[the remark this answers is not in the document]
Comment (MoKob) [reply include/extractor/node_based_edge.hpp]: Bitfields cannot be used with `std::tie`. It does not compile.

File: include/util/node_based_graph.hpp [hunk 2]
```diff
@@ -19,70 +21,61 @@ namespace util
...
+    extractor::NodeBasedEdgeClassification flags;
+    AnnotationID annotation_data;
 };
 
+// Check if two edge data elements can be compressed into a single edge (i.e. match in terms of
+// their meta-data).
+inline bool CanBeCompressed(const NodeBasedEdgeData &lhs,
+                            const NodeBasedEdgeData &rhs,
+                            const extractor::EdgeBasedNodeDataContainer &node_data_container,
+                            bool ignore_name)
```
[the remark this answers is not in the document]
Comment (MoKob) [reply include/util/node_based_graph.hpp]: That might actually be a regression. I had two different versions in the code, that's why I introduced the variable. I will have to check here.

File: include/extractor/node_based_edge.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (oxidase) [reply include/extractor/node_based_edge.hpp]: can you share you compiler version? 
with clang 3.8.0 there is no problem
http://coliru.stacked-crooked.com/a/ec4db3cd165ea930

[the remark this answers is not in the document]
Comment (MoKob) [reply include/extractor/node_based_edge.hpp]: I take everything back. Must have been a blip. I got some error about tieing to the bitfield, but when I just tried to reproduce it, it compiled just fine. 😕

File: include/util/node_based_graph.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MoKob) [reply include/util/node_based_graph.hpp]: Ok, update here: the `ignore_name` can be removed and is just a relict from refactoring `CanBeCombined` and `IsCompatibleTo` functions, which were previously used but are now referring to different data entities. (name_id is not part of the `node_based_edge` anymore but instead an annotation.)

File: src/extractor/node_based_graph_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    // remove all remaining elements
+    annotation_data.erase(new_end, annotation_data.end());
+    util::Log() << " graoh compression removed " << (old_size - annotation_data.size())
+                << " annotations of " << old_size;
+}
+
+void NodeBasedGraphFactory::ReleaseOsmNodes()
+{
+    // replace with a new vector to release old memory
+    osm_node_ids = {};
```
[the remark this answers is not in the document]
Comment (oxidase) [reply src/extractor/node_based_graph_factory.cpp]: just checked `PackedVector` has no `clear`, `shrink_to_fit` and `swap`. I think it is good to add to `detail::PackedVector` methods
```
    inline void swap(PackedVector &other) noexcept
    {
        std::swap(vec, other.vec);
        std::swap(num_elements, other.num_elements);
    }
   void clear() noexcept
  {
     vec.clear();
    num_elements = 0;
  }

void shrink_to_fit()
{
   vec.shrink_to_fit();
}
```

and use `PackedOSMIDs().swap(osm_node_ids)`or `osm_node_ids.clear(); osm_node_ids.shrink_to_fit();`

**Comment by MoKob** [comment]
Ticketed further improvements in https://github.com/Project-OSRM/osrm-backend/issues/4585

**Comment by MoKob** [comment]
From the query performance, we can only see (if any) a positive impact. I wouldn't account for major changes here, since the CH creation alone can introduce enough randomness into the process. Still, from a performance perspective, we are good here.

times-1 dragged
times-2 dragged

**Comment by MoKob** [comment]
@oxidase can you give this one a final parse, whether I addressed your concerns correctly?

**Comment by oxidase** [comment]
@MoKob looks good to me, but there is again a merge conflict because of `.flags` is  missing

**Pull request merged**